### ISSUE TITLE
dag_direct_scheduler: Direct scheduler does not support processing multiple unsubmitted nodes

### ISSUE THREAD

**Issue by al42and** [AdaptiveCpp/AdaptiveCpp#728]
**Bug summary**

When using hipSYCL runtime from multiple threads in the same process on a single CUDA GPU, occasionally the following error is thrown:

```
[hipSYCL Error] from /home/aland/hipSYCL/src/runtime/dag_direct_scheduler.cpp:313 @ submit(): dag_direct_scheduler: Direct scheduler does not support processing multiple unsubmitted nodes
============== hipSYCL error report ============== 
hipSYCL has caught the following undhandled asynchronous errors: 

   0. from /home/aland/hipSYCL/src/runtime/dag_direct_scheduler.cpp:313 @ submit(): dag_direct_scheduler: Direct scheduler does not support processing multiple unsubmitted nodes
The application will now be terminated.
```

The error is progressively more likely to occur with increasing number of threads (always happens with 9 threads, occasionally happens with four). Setting `HIPSYCL_RT_SCHEDULER=unbound` does **not** change the behavior. Setting `HIPSYCL_DEBUG_LEVEL=3` or 4 makes the error go away consistently. No warnings reported at level 2 besides the inability to load HIP plugin.

**To Reproduce**

I could not easily assemble a small reproducer, since it seems to depend on the multithreaded nature of the code and perhaps also on repeated re-initialization of the device backend with creation and destruction of SYCL queues for each test case.

Ubuntu 20.04, CUDA 11.6, hipSYCL develop (1046a78777a23ca75c6ea4e92291f1dbe36169ca):

```bash
wget https://ftp.gromacs.org/gromacs/gromacs-2022.tar.gz && tar zxf gromacs-2022.tar.gz && cd gromacs-2022/
mkdir build && cd build
cmake .. -DCMAKE_BUILD_TYPE=RelWithDebInfo -DCMAKE_C_COMPILER=clang-12 -DCMAKE_CXX_COMPILER=clang++-12 -DGMX_GPU=SYCL -DGMX_SYCL_HIPSYCL=ON -DHIPSYCL_TARGETS='cuda:sm_75'
nice cmake --build . --target tests -- -j$(nproc)
# A lot of "[hipSYCL Warning] CBS pipeline might not result in peak performance with old PM" warnings
./bin/mdrun-mpi-test -ntmpi 9
# Observe the failure. Happens in a random test case each time.
# Setting HIPSYCL_DEBUG_LEVEL=3 makes the problem go away
```

**Expected behavior**

hipSYCL runtime does not throw any errors, test passes.

**Describe your setup**
* hipSYCL develop (1046a78777a23ca75c6ea4e92291f1dbe36169ca)
* Describe the dependencies that hipSYCL sits on top of in your setup: 
   * Clang 12 from Ubuntu repos
   * Same Clang 12
   * CUDA 11.6
* How you have compiled your application and which arguments you have passed to `syclcc`. In particular, which backends and hardware you have compiled for.

**Additional context**

Unfortunately, I could not come up with a short reproducer, but perhaps the description would be enough to find the source of the bug? The issue is not critical for us, but random failures from the runtime are not nice.

**Comment by fxzjshm** [comment]
I also encountered this problem when adding multi-threading to an existing program. Several asserts/errors is ramdomly triggered during execution, and they all seems to be related to statements like `node->get_requirements()->is_submitted()`.  

A piece of code is found that triggers multi-threading problems. This triggers `assert(req->is_submitted());` in `dag_node::mark_virtually_submitted()` most, and triggered "dag_direct_scheduler: Direct scheduler does not support processing multiple unsubmitted nodes" once during my tests. (but only once, so I'm not sure whehter it is triggered by the code below. I'm not sure whether these are the same problem either, so sorry if I'm wrong.)   

Hope this helps.
```c++
#include <sycl/sycl.hpp>

int main() {
  int n_elems = 1 << 6;
  int n_threads = 2;
  sycl::device dev;
  sycl::queue q(dev);
  std::vector<sycl::buffer<int>> bufs(n_threads, sycl::range<1>(n_elems));

  for (auto i = 0; i < n_threads; i++) {
    const int j = i;
    std::thread([&, j]() {
      const int k = j;
      if (k >= n_threads) {
        return;
      }
      while (1) {
        q.submit([&](sycl::handler &cgh) {
          auto aI1 = bufs[k].get_access(cgh);
          auto aI0 = bufs[0].get_access(cgh);
          auto idx = std::rand() % n_elems;
          cgh.single_task([=]() { aI1[idx] = aI0[idx]; });
        }).wait();
        auto idx = std::rand() % n_elems;
        auto aI1 = bufs[k].get_access<sycl::access_mode::read_write>();
        aI1[idx] = 233;
        //q.wait_and_throw();
      }
    }).detach();
  }

  while (1) {
    std::this_thread::yield();
  }

  return 0;
}
```

**Comment by al42and** [comment]
Bisecting leads to #696 (in retrospect, the PR title was a giveaway).

Reverting ba9d2605a4916bc936a00ed12d595dc5b9142e0e (`git revert ba9d2605`, not necessary to revert the other commit in #696) fixes the issue with GROMACS.

**Comment by illuhad** [comment]
It seems the reproducer from @fxzjshm does not trigger any thread sanitizer reports. Don't know yet what is going on here.

**Comment by illuhad** [comment]
With #739, @fxzjshm's reproducer now longer fails for me (I've aborted execution after 30min without error).
